### PR TITLE
Differentiate between URI and SSL settings for in-cluster vs. submission

### PR DESCRIPTION
**Pull request by mccheah**
Allows users to provide separate settings for contacting the resource staging server from within the init-container vs. from the submission client. One example use case for this is that the resource staging server might be running inside the cluster, and exposed to the submission client through an Ingress or some other front door URI. In this case, the resource staging server should be accessed from the submission client via the Ingress URI, but the init-containers that are running inside the cluster should access the resource staging server through the pod IP or service IP. The certificate that has to be presented when contacting over the external URI may also differ from the certificate that needs to be used when contacting through the internal URI.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:376]: with a scheme like `file://` or without?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: file -> file://

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:416]: what scheme?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: this .orElse isn't aligned with the others

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: can you not use the requireNand from elsewhere here?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: file:// and local://

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala]: file:// and local://

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: was this not being used before?  I'm surprised checkstyle or something didn't alert that there was an unused identifier

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala]: where is this called from?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:376]: "Strictly a path" I think implies no scheme.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:416]: "Strictly be a path" denotes no scheme.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/InitContainerResourceStagingServerSslOptionsProvider.scala]: I think this was an original idea I had but abandoned it later, forgot to remove the class.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: Yeah unfortunately this slipped under the radar somehow. Are the SBT tests running now? Maybe SBT would have caught it and Maven doesn't?

Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: they're not running right now, but I can temporarily re-enable, you push a commit, and we watch them run to see if they would've caught this

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:326]: No need specifically but I suppose it's all the more reason to look into getting our SBT builds online.

**Comment by mccheah** [comment]
Requires rebasing onto allow-client-cert-pem

**Comment by mccheah** [comment]
Finished rebasing here for now.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/SubmittedDependencyInitContainerConfigPlugin.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/SubmittedDependencyInitContainerConfigPlugin.scala:90]: maybe in a fixup commit after this merges can add "has invalid scheme for secret" to give end user just a bit more context

**Review by ash211 — APPROVED** [review]
LGTM, planning to merge when this integration test passes

**Pull request merged**
